### ISSUE TITLE
Does it support CAN? Any english doc available?

### ISSUE THREAD

**Issue by parthi2929** [toppers/athrill#62]
Hi

This RH850 looks great, can you let us know if it supports CAN, and if english detailed manuals available on how to use it?

**Comment by parthi2929** [comment]
We are interested in how to run TOPPER ATK2, RTE, BSW esp comstack on this simulator, between 2 virtual CPUs.

**Comment by tmori** [comment]
yes, RH850 athrill supports CAN. 

CAN device sources:
https://github.com/toppers/athrill-target-rh850f1x/tree/master/src/device/peripheral/rh850f1k/can

TOPPERS ATK2/RTE/comstack also can be executed on RH850 athrill. But, public documents does not exist...
So you need to integrate the TOPPERS/AUTOSAR(ATK2) stack yourself.

**Comment by parthi2929** [comment]
I am depicting my understanding as below. 

<img src="https://user-images.githubusercontent.com/11896263/177676121-8bca52fb-f1ad-4fbe-aeb7-cee07f39e768.png" width=450px>

1. So today I could do the case A above. Please confirm if my understanding is correct here. 
2. If correct, is this simulation just to see if compilation produces executable file for CPU successfully? or we can run and see, for ex, the simulated CPU registers? 
3. If correct, then our requirement is in case B. This we should stack ourselves, and you just confirmed, can run [ATK2](https://www.toppers.jp/atk2.html), [A-COMSTACK](https://www.toppers.jp/a-comstack.html), [A-RTEGEN](https://www.toppers.jp/a-rtegen.html). As you said public documents not yet existing, can you please point to who has them, then we could probably request them to share? 
4. We basically want to simulate below minimalistic scenario with simple IO and CAN 

    <img src="https://user-images.githubusercontent.com/11896263/177678332-2daf0d94-6cfd-48a1-943d-98b26d18ff1a.png" width=650px>

    Can we do that, by some kind of simulated arrangement like below with your target? Here, for ex, program A from a virtual RH850 CPU can send a CAN, program B running on another virtual RH850 CPU can receive CAN, but can a virtual CAN communication be established between them to exchange CAN messages like this? And can an external software (a typical CAN bus analyzer) can also probe this _virtual CAN_? 

    <img src="https://user-images.githubusercontent.com/11896263/177678520-48597089-6b91-445a-b4bd-979b07bcb816.png" width=650px>

**Comment by tmori** [comment]
> 1. So today I could do the case A above with your simulator out of the box. Please confirm if my understanding is correct here.
yes. It is correct with your understanding as shown in the picture case A.

**Comment by tmori** [comment]
> 2. If correct, is this simulation just to see if compilation produces executable file for CPU successfully? or we can run and see, for ex, the simulated CPU registers?

Yes, not only can you run the executable, but you can also debug it to check CPU registers and memory.

The Athrill manual is available at the following site.

https://qiita-com.translate.goog/kanetugu2018/items/cf3dea16710a3f0737e8?_x_tr_sl=ja&_x_tr_tl=en&_x_tr_hl=ja&_x_tr_pto=wapp

**Comment by tmori** [comment]
> If correct, then our requirement is in case B. This we should stack ourselves, and you just confirmed, can run [ATK2](https://www.toppers.jp/atk2.html), [A-COMSTACK](https://www.toppers.jp/a-comstack.html), [A-RTEGEN](https://www.toppers.jp/a-rtegen.html). As you said public documents not yet existing, can you please point to who has them, then we could probably request them to share?

@mitsut  Could you provide your suggestion about Q3. I think you are the best guy to answer it.

**Comment by tmori** [comment]
> 4. We basically want to simulate below minimalistic scenario with simple IO and CAN
> Can we do that, by some kind of simulated arrangement like below with your target? Here, for ex, program A from a virtual RH850 CPU can send a CAN, program B running on another virtual RH850 CPU can receive CAN, but can a virtual CAN communication be established between them to exchange CAN messages like this? And can an external software (a typical CAN bus analyzer) can also probe this virtual CAN?

The first question is Yes. We can communicate via CAN protocol among several Athrill instances.
The second question is unfortunately No. The current implementation for CAN communication on Athrill is just an emulation, which means CAN protocol will be converted to [ROS](https://wiki.ros.org/) communication. So we don’t think you can observe the internal of CAN communication from an external CAN bus analyzer.

IOW, you can observe data on the CAN bus as the ROS topic (e.g., ros topic echo <CANID>). We wish you are familiar with ROS.

**Comment by parthi2929** [comment]
I tried to grasp ROS to visualize this situation better, but could not go much (its becoming another big topic to study 🙂)

From my limited understanding, I conjecture that something like below could be possible, where if the emulator instances are communicating with ROS messages, we could tap that, convert back to CAN and then a custom CAN Bus analyzer can show that as raw CAN messages? 

![image](https://user-images.githubusercontent.com/11896263/177756851-ebe58676-1e60-421c-844b-3a97a0501135.png)

**Comment by tmori** [comment]
> I conjecture that something like below could be possible, where if the emulator instances are communicating with ROS messages, we could tap that, convert back to CAN and then a custom CAN Bus analyzer can show that as raw CAN messages? 
> The pink dotted line simply implies an equivalence of virtual CAN bus realization, but under the hood, ROS internal communication happening indicated by black line.

Excellent! Yes, it is possible to monitor CAN data by adding the conversion mechanism on the right side of the diagram depicted.

Also, by sending test data from the monitor side to the virtual CAN bus, it would be possible to check the operation of Program A/B.

**Comment by parthi2929** [comment]
Great, I guess I could proceed as below. 

1. First PoC to successfully run 2 RH850 Ahtril instances in our PC, and make them communicate via CAN Tx/Rx
![image](https://user-images.githubusercontent.com/11896263/178137101-564d3280-ef0f-4463-ab7f-adbcdec9b495.png)

1. Then a custom built raw ROS Reader node, to read the raw CAN data
![image](https://user-images.githubusercontent.com/11896263/178137124-e03821d2-dfb9-42ff-b392-211229315362.png)

Some challenges ahead. 
1. We do not have any ROS expertise yet (for that matter, TOPPER stack also is something next to be learnt and get acquainted with). Any directions on docs that could specifically help here, like Athril ROS communication documentation, could be very helpful, so we could focus on untangling that part.   
2. I also believe, such a CAN monitor for Athril, could highly benefit the public as well, so its also a good fit for a funded program from universities, interested parties, or volunteered/sponsored by individuals. Any suggestions on how we could pursue that could be welcome (that CAN monitor could be part of this repo or separate available to public).   
4. Even if not a full blown Bus analyzer, a virtual drive simulating CAN Tx/Rx from ROS, for which a 3rd party CAN bus analyzer like [PCAN Viewer](https://www.peak-system.com/PCAN-View.242.0.html?&L=1) can use, could be very helpful. 
5. RH850 boards are 4 or 5 times as costly as other mainstream automotive MCUs from Infineon, STMicroelectronics, and NXP. Would be grateful if we get ATK simulators for Infineon Aurix TC3XX series, or NXP MPC5xxx  or ST's SPC56 family. If in any way possible, please support internally in your community.

**Comment by tmori** [comment]
@parthi2929 Great!

 If environment 1 can be constructed, realization of 2 will be easy.

However, as you are concerned, the hurdle to building an environment may be high due to the lack of documentation to refer to. Therefore, we have created the following repository and are preparing it so that you can set up a basic environment.

https://github.com/toppers/hakoniwa-ecu-multiplay

In this repository, the environment is built on a docker basis, and you can refer to the Dockerfile to understand the environment building procedure.

At this time, the ROS environment, the RH850 athrill, and the gcc compiler have already been installed.

Since the environment is built with docker-compose, multiple instances can be launched on the virtual environment.

**Comment by tmori** [comment]
@parthi2929 

The following article describes a sample program for visualising CAN communication in ROS2.
I hope this will be of some help.

https://qiita.com/kanetugu2018/items/83f838780f88078386c9

**Comment by parthi2929** [comment]
Thank you very much @tmori . For past few weeks, could not return on this topic, but I expect in a month or two. I am very grateful for your support on this, and very excited about our future collaboration on this.

**Comment by Subham2372** [comment]
Hello @tmori, 
Hope you are all good. I was going through the Comstack in topper stack of yours. 
and i have question related to that. 
There are many bat file which needed to be run for generating the file but whenever i am clicking the bat to run its open and close. 
Do you have any precondition like any Software or any particular path to be set. 
as per the guidelines i have set the path but i am not able to see anything there generating. 
Could you please guide me here with that. 
Thank you in advance!!!

**Comment by mitsut** [comment]
Hi, @Subham2372 

Thank you for using a-comstack.
Do you have a question about which is the file a-comstack/can/target/hsbrh850f1k_ccrh/sample/configure.bat?
This bat file is the process to generate the c-files needed to build a-comstack from the ARXML file.

If you try to run this bat file, you need cfg.exe is required.
https://www.toppers.jp/download.cgi/cfg-mingw-static-1_9_6.zip (for Windows)

If you need a more detailed answer, please feel free to start a new thread in the discussion.
https://github.com/toppers/users-forum/discussions/categories/q-a

Thank you.
